osbs/utils.py

```diff
@@ -190,72 +191,122 @@ def buildconfig_update(orig, new, remove_nonexistent_keys=False):
-                    logger.info("retrying command '%s' in %ss:\n '%s'", args, backoff, ex.output)
-                else:
-                    logger.info("retrying command '%s':\n '%s'", args, ex.output)
-                time.sleep(backoff)
+            # we are using check_output, even though we aren't using
+            # the return value, but we will get 'output' in exception
+            subprocess.check_output(cmd, stderr=subprocess.STDOUT)
+            repo_commit, repo_depth = reset_git_repo(target_dir, commit, depth)
+            break
+        except subprocess.CalledProcessError as exc:
+            if counter != retry_times:
+                logger.info("retrying command '%s':\n '%s'", cmd, exc.output)
+                time.sleep(retry_delay * (2 ** counter))
             else:
-                raise
+                raise OsbsException("Unable to clone git repo '%s' "
+                                    "branch '%s'" % (git_url, branch),
+                                    cause=exc, traceback=sys.exc_info()[2])
 
+    return ClonedRepoData(target_dir, repo_commit, repo_depth)
 
-@contextlib.contextmanager
-def checkout_git_repo(git_uri, git_ref, git_branch=None, max_retries=GIT_MAX_RETRIES,
-                      backoff_factor=GIT_BACKOFF_FACTOR):
-    """
-    Context manager to check out a git repository, yielding a string with the path
-    for the cloned repository.
-
-    :param git_uri: string, git repository URI
-    :param git_ref: string, target git reference
-    :param git_branch: string, target git branch
-    :param max_retries: int, maximum number of git clone retries
-    :param backoff_factor: int, exponential backoff factor for retries, in seconds
+
+def reset_git_repo(target_dir, git_reference, retry_depth=None):
     """
-    tmpdir = tempfile.mkdtemp()
-    repo_path = os.path.join(tmpdir, "repo")
-    # when you clone into an empty directory and cloning process fails
-    # git will remove the empty directory (why?!)
-    args = ['git', 'clone', git_uri]
-    if git_branch:
-        args += ['-b', git_branch]
-
-    args.append(repo_path)
-    try:
-        try:
-            subprocess_check_output_with_retry(args, max_retries, backoff_factor)
-        except subprocess.CalledProcessError as ex:
-            raise OsbsException("Unable to clone git repo '%s' "
-                                "branch '%s'" % (git_uri, git_branch),
-                                cause=ex, traceback=sys.exc_info()[2])
+    hard reset git clone in target_dir to given git_reference
 
-        # Find the specific ref we want
+    :param target_dir: str, filesystem path where the repo is cloned
+    :param git_reference: str, any valid git reference
+    :param retry_depth: int, if the repo was cloned with --shallow, this is the expected
+                        depth of the commit
+    :return: str, commit ID of HEAD
+    """
+    deepen = retry_depth or 0
+    commit_depth = None
+    for _ in range(1, 10):
```

Comment:
Is there any reason for using this specific range here? Could we use just `range(9)` and set it as a constant in the constant modules?

Reply:
No. Yes. Fixed.